### PR TITLE
Allow to select the instance and the main method to execute when launching the interpreter (#60, #86)

### LINKED ISSUE #60: No ALE Run Configuration created when launching ALE through the contextual menu shortcut
## Expected behavior

When ALE is launched through the contextual menu shortcut a new _ALE launch_ Run Configuration should be created. The main benefit is that we could easily run a same script multiple times in a row by just typing Ctrl+F11 or searching a previously launched script through the Run History.

> Note: what I mean by "contextual menu shortcut" is: 
> - _Right clicking on a .dsl file > Run As > ALE launch_

## Current behavior

No Run Configuration is created. Hence, we have use the contextual menu each time we want to launch the same script or manually create a dedicated Run Configuration.

## Solution

To solve this, the usual algorithm is (within a launch shortcut instance):
1. Retrieve the template of the desired Launch Configuration
2. Create a new Launch Configuration instance based on the template
3. Set the Launch Configuration's attributes
4. Run the Launch Configuration

For example:
```java
public class LaunchShortcut implements ILaunchShortcut {

    @Override
    void launch(ISelection selection, String mode) {
        ILaunchManager manager = DebugPlugin.getDefault().getLaunchManager();
        ILaunchConfigurationType type = manager.getLaunchConfigurationType(AleRunConfiguration.ID);
        ILaunchConfiguration[] configurations = manager.getLaunchConfigurations(type);

        // Create a new launch configuration

        String baseConfigurationName = configurationNameFor(dslFileName);
        String launchConfigurationName = availableLaunchConfigurationName(baseConfigurationName, configurations);
	
        ILaunchConfigurationWorkingCopy workingCopy = type.newInstance(null, launchConfigurationName);        
        workingCopy.setAttribute(AleRunConfiguration.MODEL_FILE, ...);
        workingCopy.setAttribute(AleRunConfiguration.DSL_FILE, ...);
        ILaunchConfiguration configuration = workingCopy.doSave();
		
        // Run the new launch configuration
        configuration.launch(ILaunchManager.RUN_MODE, null);
    }

}
```

### LINKED ISSUE #86: Ability to select the instance and the main method to execute when launching the interpreter
## Current Behavior

When launching an ALE script we provide two information:

 - the DSL file (_.dsl_) providing the metamodel and its semantics
 - the model file (_.xmi_) defining the model on which the semantics should be applied

Currently, ALE's launcher:

1. Parse the model
2. Find the first method annotated with `@main`
3. Call this method on the model's root element

## Issues

This behavior causes the following problems:

- if the root element is not of the type expected by the `@main` method ALE doesn't show any error, which can confuse the user
- if several `@main` methods are available ALE will arbitrarily select the first one, whereas the user may want to run another one
- it is impossible to run a `@main` method on a model element which is not at the root of the XMI file

## Expected Behavior

At runtime...

### ... when there are no ambiguity

If the semantics provides only one `@main` method and the model contains only one element that supports the semantics then this method should be run automatically on the corresponding element.

### ... where there is an ambiguity

If the semantics provides several `@main` methods or if the model contains several elements that support the semantics then the Launch Delegate should prompt a dialog to ask the user to clarify their will.

A similar behavior [is provided by the GEMOC Studio](http://gemoc.org/studio_docs/2.3.1/Guide.html#modeling-workbench-executing-model-section).

## Related Issue

#60 — Create a Run Configuration when the context-menu shortcut is used to run the interpreter

### PR DESCRIPTION
**Pull request by echebbi**
### Contribution

Fix #60 and fix #86 by:
 - improving the Launch Configuration tab to allow a user to select a `@main` method and the element of the model on which it should be called
 - asking the user to select a couple (method / element) when several `@main` methods / target elements are available and the user did not specified the ones to use
 - persisting the launch configuration created by the contextual menu (right-click > Run As) so that it can thereafter be run with CTRL+F11 (on Windows).

### Comments

I did not implemented new automatic tests, but I manually tested various scenarios and it seems to work as expected.

This PR is a draft because I'd like to improve the user documentation before merging.

#### The new Launch Configuration tab

<div align="center">

</div>

#### The main method selection dialog:

<div align="center">

</div>

#### The main model element selection dialog:

<div align="center">

</div>

#### The entry point selection dialog:

(called when ALE is executed but an ambiguity is detected)

<div align="center">

</div>

> _Note:_ the choise is persisted so that the user won't have to make it again for this launch configuration

### CODE REVIEW

File: plugins/org.eclipse.emf.ecoretools.ale.ide.ui/src/org/eclipse/emf/ecoretools/ale/ide/ui/launchconfig/UiUtils.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+/*******************************************************************************
```
Comment (dvojtise) [inline plugins/org.eclipse.emf.ecoretools.ale.ide.ui/src/org/eclipse/emf/ecoretools/ale/ide/ui/launchconfig/UiUtils.java]: This class looks useful, maybe consider having a copy in GEMOC commons too

File: plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/parser/internal/DatedDslSemantics.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+import org.eclipse.emf.ecoretools.ale.implementation.Method;
+import org.eclipse.emf.ecoretools.ale.implementation.ModelUnit;
+
+/**
+ * The semantics of a DSL at some point in time.
+ * <p>
+ * An instance of this class is not updated as the underlying resource
+ * defining the semantics (e.g. a <i>.dsl</i> or a <i>.ale</i> file) changes.
+ */
+public class DatedDslSemantics implements DslSemantics {
```
Comment (dvojtise) [inline plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/parser/internal/DatedDslSemantics.java]: Maybe the name of this class is misleading
Date implies to have a date information in the class but there is not
 while the purpose of the class seems to have an immutable DslSemantic ?

**Comment by echebbi** [comment]
I updated the "entry point selection dialog" so that it now displays the name of the EObject in addition to its fully qualified name (see first comment for the screenshot).

I also renamed `DatedDslSemantics` as `ImmutableDslSemantics`.

Currently a lot of tests do not pass; that's because the `ALEInterpreter` does not automatically uses semantics' first `@main` method and model's first element anymore. Indeed, the method to call and the model element on which it will be called must be given as arguments (see [ALEInterpreter#L241](https://github.com/gemoc/ale-lang/pull/89/files#diff-165f5f16a5e9f2a8aa366838c2f5730b)). 

Ideally I'd like to keep it that way because that's not the business of the interpreter to make such choices and that's why I gave this reponsibility to `AleLaunchConfigurationDelegate`. I'm going to think about a solution, maybe based on the Strategy pattern to allow different behaviors (e.g. use first main method by default, or cancel launch when ambiguous, or prompt a dialog to ask for a selection). Anyhow, this will be part of ALE's public API so it must be easy to use.

File: plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/parser/internal/DatedDslSemantics.java (hunk 2 above)
Comment (echebbi) [reply plugins/org.eclipse.emf.ecoretools.ale.core/src/org/eclipse/emf/ecoretools/ale/core/parser/internal/DatedDslSemantics.java]: Renamed as `ImmutableDslSemantics`

File: plugins/org.eclipse.emf.ecoretools.ale.ide.ui/src/org/eclipse/emf/ecoretools/ale/ide/ui/launchconfig/UiUtils.java (hunk 1 above)
Comment (echebbi) [reply plugins/org.eclipse.emf.ecoretools.ale.ide.ui/src/org/eclipse/emf/ecoretools/ale/ide/ui/launchconfig/UiUtils.java]: In the [org.eclipse.gemoc.commons.eclipse.ui](https://github.com/eclipse/gemoc-studio/tree/f413e861a80678eb9d4f55268d032f9c31f90f6a/commons/plugins/org.eclipse.gemoc.commons.eclipse.ui) bundle?

**Comment by echebbi** [comment]
The [tests now pass](https://ci.inria.fr/gemoc/job/ale-lang/job/enhance-launch-configuration/), so this PR can be merged.

### Important: introduces breaking changes:

- The only method to execute ALE's interpreter is now `ALEInterpreter.eval(EObject, Method, List<Object, DslSemantics)`

Although worth mentioning this change should be easy take over; see changes made to `examples/logo.standalone/src/logo/standalone/Main.java` for an example.

### Note for later

`ALEInterpreter.eval` fails when called with a `caller` that has been loaded in another EMF context (see JavaDoc for details). I think that it shows a lack of cohesion and that we should come up with a better design. Maybe update interpreter's `queryEnvironment` at the beginning of the execution with relevant EPackages rather than relying on the interpreter to create the semantics and load the model.

**Pull request merged**